### ISSUE TITLE
Workflow with `watchPath` does not finish on v22.10.4 even though `until` condition was met

### ISSUE THREAD

**Issue by julibeg** [nextflow-io/nextflow#3497]
### Expected behavior and actual behavior

When creating a channel with `watchPath` on v22.10.4, nextflow does not terminate at the end of the workflow (even though a file matching the `until` condition was found)

### Steps to reproduce the problem

create a file `test.nf`
```
dir = "watchPath-test-dir"

workflow {
  ch = Channel.watchPath("$dir/*").until{ it.name == "done" }
  ch.subscribe \
    onNext: { print "emitted: $it.name" },
    onComplete: { print "we're done" }

  ch.collect() | view { "emitted: ${it.size()} files in total" }
}
```
Then, create a directory called `watchPath-test-dir`. Run `nextflow run test.nf` and (in a different terminal) `touch` files in `watchPath-test-dir`. When `touch`ing a file called `done`, the channel finishes and the rest of the workflow is executed. At the very end, however, nextflow doesn't exit.

### Program output 
```
N E X T F L O W  ~  version 22.10.4
Launching `test.nf` [naughty_ardinghelli] DSL2 - revision: 2c89a0ad4f
emitted: file1
emitted: file2
we're done
emitted: 2 files in total

// hangs at this point
```

`.nextflow.log` even has `Dec-15 12:24:40.488 [main] DEBUG nextflow.script.ScriptRunner - > Execution complete -- Goodbye` as its last line.

### Environment 

* Nextflow version: 22.10.4
* Java version: openjdk 11.0.13 2021-10-19
* Operating system: Ubuntu
* Bash version: GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)

### Additional context

When trying the same with v22.04.0, it exits as expected.

**Comment by SamStudio8** [comment]
As this works in 22.04 and not 22.10, I checked the CHANGELOG for relevant changes. I spotted this in 22.06:

```
22.06.0-edge - 9 Jun 2022
[...]
- Add DirWatcher v2 [209c82cd]
```

I tried @julibeg's example above and indeed it works as expected in 22.05 and hangs indefinitely in 22.06.

This [change](https://github.com/nextflow-io/nextflow/commit/209c82cd) introduced DirWatcherV2 and made it the default. We can further confirm the new behaviour is at fault by using the [NXF_DIRWATCHER_LEGACY](https://github.com/nextflow-io/nextflow/commit/209c82cd#diff-8bb0cdd5d574d3f5a2b446633cd0d1755470d396713ba151cb7f8bba7e22f106R312) environment variable that switches back to the original DirWatcher. The example above exits as expected with `NXF_DIRWATCHER_LEGACY=true` on 22.10.0.

As @julibeg indicated, the execution of the workflow does finish, so I suspect something is not released when the watcher is [terminated](https://github.com/nextflow-io/nextflow/commit/209c82cd#diff-46f579c9b7822a9200172c99c09eae887700a357955e7d042c650576b00a3ea9R137).

**Comment by bentsherman** [comment]
I was able to reproduce the issue, and what I find interesting is that the `terminate()` method is not called for DirWatcher v1 or v2. I added a log message to both methods and didn't see it in the log in either case.

When watchPath is called, it creates the DirWatcher like so: https://github.com/nextflow-io/nextflow/blob/d1bb2fe2906303c6c5386e47136b7d5d104b7fec/modules/nextflow/src/main/groovy/nextflow/Channel.groovy#L309-L328

Notice that the watcher itself is not preserved anywhere. @SamStudio8 to your point about the UntilOp calling `proc.terminate()`, I think that is the watcher's output channel but not the watcher itself.
